DM-42187: Add RemoteButler.getURIs

dhirving opened the pull request:
## Checklist

- [ ] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`

dhirving committed:
Implement RemoteButler.getURIs

This re-uses the existing server endpoints for getting file info.

dhirving committed:
Implement RemoteButler.getURI

Moved the DirectButler implementation to the Butler base class, since the code is identical for RemoteButler.

timj commented on python/lsst/daf/butler/remote_butler/_remote_butler.py:
```diff
@@ -287,9 +289,26 @@ def getURI(
         # Docstring inherited.
-        raise NotImplementedError()
+        if predict or run:
+            raise NotImplementedError("Predict mode is not supported by RemoteButler")
```
We need to discuss whether we can make this work. Obviously a signed URL to a resource that doesn't exist is not all that helpful but people like to know what the filename might look like. We can tell them that in theory -- maybe for predict mode we return the unsigned `s3` URI? Can punt this to another ticket.

timj commented on tests/test_server.py:
```diff
@@ -104,9 +106,11 @@ def setUpClass(cls):
-
         # Add a file with corrupted data for testing error conditions
         cls.dataset_with_corrupted_data = _create_corrupted_dataset(cls.repo)
+        # All of the datasets that come with MetricTestRepo are disassembled
```
I had forgotten that the MetricTestRepo defaults to storage class `StructuredCompositeReadComp`.  Easy to change the default if we wanted to but also nice to know that getURIs should work.

timj commented on tests/test_server.py:
```diff
@@ -323,6 +328,50 @@ def check_sc_override(converted):
+        with self.assertRaises(RuntimeError):
+            self.butler.getURI(dataset_type, dataId=data_id, collections=collections)
+
+        # getURIs does NOT respect component overrides on the DatasetRef,
```
There is only one URI for any of the components if not disassembled. I imagine no-one has ever tried to request the URI of an individual component by passing in the ref but it seems like it should return the one URI of the component.

timj reviewed (approved):
Looks good.

dhirving committed:
Pin documenteer

The newly-released Documenteer 1.0 requires additional configuration and does not support the pipelines.lsst.io theme yet.
